### PR TITLE
Implement `FileHandle` for asset file system

### PR DESCRIPTION
**Pull request by JakeWharton**
While uncompressed assets can be accessed using a file descriptor and normal random access methods, most assets will be compressed and thus can only be read linearly.

### CODE REVIEW

File: okio-assetfilesystem/src/androidTest/kotlin/okio/assetfilesystem/AssetFileSystemTest.kt [hunk 1]
```diff
@@ -175,6 +179,79 @@ class AssetFileSystemTest {
...
+
+    val mobySize = fs.openReadOnly("moby10b.txt".toPath()).use(FileHandle::size)
+    assertThat(mobySize).isEqualTo(1232923)
+  }
+
+  @Test fun openReadOnlyRandomAccessForward() {
+    fs.openReadOnly("moby10b.txt".toPath()).use { handle ->
+      val buffer = Buffer()
+      handle.read(34251, buffer, 16)
+      assertThat(buffer.readUtf8()).isEqualTo("Call me Ishmael.")
```
Comment (swankjesse) [inline okio-assetfilesystem/src/androidTest/kotlin/okio/assetfilesystem/AssetFileSystemTest.kt:206]: YASSSSS!

File: okio-assetfilesystem/src/androidTest/kotlin/okio/assetfilesystem/AssetFileSystemTest.kt [hunk 2]
```diff
@@ -175,6 +179,79 @@ class AssetFileSystemTest {
...
+      handle.read(148605, buffer, 49)
+      assertThat(buffer.readUtf8()).isEqualTo("It is not down in any map; true places never are.")
+      handle.read(1051694, buffer, 50)
+      assertThat(buffer.readUtf8()).isEqualTo("his forehead's veins swelled like overladen brooks")
+      handle.read(148605, buffer, 49)
+      assertThat(buffer.readUtf8()).isEqualTo("It is not down in any map; true places never are.")
+      handle.read(34251, buffer, 16)
+      assertThat(buffer.readUtf8()).isEqualTo("Call me Ishmael.")
```
Comment (swankjesse) [inline okio-assetfilesystem/src/androidTest/kotlin/okio/assetfilesystem/AssetFileSystemTest.kt:214]: # 👨🏻‍🍳 👌🏻

File: okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt [hunk 3]
```diff
@@ -154,3 +157,84 @@ private class AssetFileSystem(
...
+  private var size = -1
+
+  override fun protectedRead(
+    fileOffset: Long,
+    array: ByteArray,
+    arrayOffset: Int,
+    byteCount: Int,
+  ): Int {
+    // If we need to jump backwards or have reached the end of the file,
+    // close the existing stream and open a new one.
```
Comment (swankjesse) [inline okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt:176]: Clever!

File: okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt [hunk 4]
```diff
@@ -154,3 +157,84 @@ private class AssetFileSystem(
...
+    } else {
+      currentOffset += read
+    }
+    return read
+  }
+
+  override fun protectedSize(): Long {
+    if (size == -1) {
+      while (true) {
+        val skipped = inputStream.skip(128).toInt()
```
Comment (swankjesse) [inline okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt:209]:
```suggestion
        val skipped = inputStream.skip(1024 * 1024).toInt()
```

We’re not here to fuck spiders.

**Review by swankjesse — APPROVED** [review]

File: okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt (hunk 4 above)
Comment (JakeWharton) [reply okio-assetfilesystem/src/main/kotlin/okio/assetfilesystem/AssetFileSystem.kt:209]: Hmm I don't remember that quote from Moby Dick

File: okio-assetfilesystem/src/androidTest/kotlin/okio/assetfilesystem/AssetFileSystemTest.kt (hunk 1 above)
Comment (oldergod) [reply okio-assetfilesystem/src/androidTest/kotlin/okio/assetfilesystem/AssetFileSystemTest.kt:206]: gotta read it down

**Pull request merged**